lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs

```diff
@@ -3338,19 +3397,41 @@ submitSharedTransaction ctx apiw@(ApiT wid) apitx = do
     void $ withWorkerCtx ctx wid liftE liftE $ \wrk -> do
 
         (cp, _, _) <- liftHandler $ withExceptT ErrSubmitTransactionNoSuchWallet $
-            W.readWallet @_ @s @k wrk wid
-        let (ScriptTemplate _ script) = (Shared.paymentTemplate $ getState cp)
-        let witsPerInput = Shared.estimateMinWitnessRequiredPerInput script
+            W.readWallet @_ wrk wid
+        let witCountCtx = toWitnessCountCtx @(SharedState n SharedKey) (getState cp)
+        let (tx,_,_,_,_, (WitnessCount _ nativeScripts _)) =
+                decodeTx tl era witCountCtx sealedTx
+
+        let numberStakingNativeScripts =
+                length $ filter hasDelegationKeyHash nativeScripts
+
+        let delegationScriptTemplateM = Shared.delegationTemplate $ getState cp
+        let delegationWitsRequired = case delegationScriptTemplateM of
+                Nothing -> 0
+                Just (ScriptTemplate _ scriptD) ->
+                    if numberStakingNativeScripts == 0 then
+                        0
+                    else if numberStakingNativeScripts == 1 then
+                        Shared.estimateMinWitnessRequiredPerInput scriptD
+                    else
+                        error "wallet supports transactions with 0 or 1 staking script"
+
+        let (ScriptTemplate _ scriptP) = Shared.paymentTemplate $ getState cp
+        let pWitsPerInput = Shared.estimateMinWitnessRequiredPerInput scriptP
 
         let witsRequiredForInputs =
                 length $ L.nubBy samePaymentKey $
                 filter isInpOurs $
                 (apiDecoded ^. #inputs) ++ (apiDecoded ^. #collateral)
         let totalNumberOfWits = length $ getSealedTxWitnesses sealedTx
-        let allWitsRequired = fromIntegral witsPerInput * witsRequiredForInputs
+        let paymentWitsRequired =
+                fromIntegral pWitsPerInput * witsRequiredForInputs
+        let allWitsRequired =
+                paymentWitsRequired + fromIntegral delegationWitsRequired
         when (allWitsRequired > totalNumberOfWits) $
             liftHandler $ throwE $
-            ErrSubmitTransactionPartiallySignedOrNoSignedTx allWitsRequired totalNumberOfWits
+            ErrSubmitTransactionPartiallySignedOrNoSignedTx
+            allWitsRequired totalNumberOfWits
```

Comment:
> Actually, why not reuse estimateNumberOfWitnesses already today

Probably because it requires all inputs to be present in the `UTxO` which we can't guarantee here. So probably not a good idea for now.